File: bindings/typescript/src/grants.ts
/**
 * The sharing (grants) views — every grant on an asset, fold-classified
 * root/delegated + active/revoked, as enumerated by `ListAssetGrants` (UI-3). Kept
 * in its own module so `types.ts` stays a thin aggregator, mirroring the Rust core's
 * module-per-concern discipline. A VIEW-only surface; issuing/revoking grants is cloud.
 */

import type {
  GrantView as PbGrantView,
  ListAssetGrantsResponse as PbListAssetGrantsResponse,
} from "./gen/kortecx/v1/gateway_pb.js";

/** One grant on an asset, fold-classified. */
export class GrantView {
  constructor(
    readonly grantor: string,
    readonly grantee: string,
    readonly actions: readonly string[],
    readonly runtimeScope: string,
    /** A root grant (from the asset owner), vs a delegated sub-grant. */
    readonly isRoot: boolean,
    /** An authorized revocation makes the grant inert in the fold. */
    readonly revoked: boolean,
  ) {}

  static fromProto(g: PbGrantView): GrantView {
    return new GrantView(g.grantor, g.grantee, g.actions, g.runtimeScope, g.isRoot, g.revoked);
  }

  /** A stable display status for the inspector. */
  get status(): "revoked" | "root" | "delegated" {
    if (this.revoked) {
      return "revoked";
    }
    return this.isRoot ? "root" : "delegated";
  }
}

/** Every grant on one asset, with the bound owner echoed (`""` if unbound). */
export class AssetGrants {
  constructor(
    readonly owner: string,
    readonly grants: readonly GrantView[],
  ) {}

  static fromProto(r: PbListAssetGrantsResponse): AssetGrants {
    return new AssetGrants(
      r.owner,
      r.grants.map((g) => GrantView.fromProto(g)),
    );
  }
}

File: bindings/python/src/kortecx/grants.py
"""UI-3 sharing (grants) views — every grant on an asset, fold-classified
root/delegated + active/revoked (``ListAssetGrants``).

Kept in its own module so ``types.py`` stays a thin aggregator. A VIEW-only surface;
issuing/revoking grants is cloud.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import List

from .v1 import gateway_pb2 as _g


@dataclass(frozen=True)
class GrantView:
    """One grant on an asset, fold-classified."""

    grantor: str
    grantee: str
    actions: List[str]
    runtime_scope: str
    is_root: bool  # a root grant (from the asset owner), vs a delegated sub-grant
    revoked: bool  # an authorized revocation makes the grant inert in the fold

    @classmethod
    def from_proto(cls, g: "_g.GrantView") -> "GrantView":
        return cls(
            grantor=g.grantor,
            grantee=g.grantee,
            actions=list(g.actions),
            runtime_scope=g.runtime_scope,
            is_root=g.is_root,
            revoked=g.revoked,
        )

    @property
    def status(self) -> str:
        """A stable display status: ``"revoked"`` | ``"root"`` | ``"delegated"``."""
        if self.revoked:
            return "revoked"
        return "root" if self.is_root else "delegated"


@dataclass(frozen=True)
class AssetGrants:
    """Every grant on one asset, with the bound owner echoed (``""`` if unbound)."""

    owner: str
    grants: List[GrantView]

    @classmethod
    def from_proto(cls, r: "_g.ListAssetGrantsResponse") -> "AssetGrants":
        return cls(
            owner=r.owner,
            grants=[GrantView.from_proto(g) for g in r.grants],
        )
